pkg/kubelet/config/file_test.go

```diff
@@ -127,109 +73,152 @@ func TestReadManifestFromFile(t *testing.T) {
+				},
+				Spec: api.PodSpec{Containers: []api.Container{{Image: "test/image"}}},
+			}),
+		},
+		{
+			desc: "Pod without ID",
+			fileContents: `{
+					"kind": "Pod",
+					"apiversion": "v1beta1",
+					"uid": "12345",
+					"desiredState": {
+						"manifest": {
+							"containers": [{ "name": "image", "image": "test/image" }],
+						},
+					},
+				}`,
+			expected: CreatePodUpdate(kubelet.SET, kubelet.FileSource, api.Pod{
+				ObjectMeta: api.ObjectMeta{
+					Name:      "12345-" + hostname,
+					UID:       "12345",
+					Namespace: kubelet.NamespaceDefault,
+					SelfLink:  "/api/v1beta2/pods/12345-" + hostname + "?namespace=default",
+				},
+				Spec: api.PodSpec{Containers: []api.Container{{Image: "test/image"}}},
+			}),
+		},
+		{
+			desc: "Pod v1beta3",
+			fileContents: `{
+					"kind": "Pod",
+					"apiversion": "v1beta3",
+					"metadata": {
+						"uid": "12345",
+						"name": "test",
+					},
+					"spec": {
+						"containers": [{ "name": "image", "image": "test/image" }],
+					},
+				}`,
+			expected: CreatePodUpdate(kubelet.SET, kubelet.FileSource, api.Pod{
+				ObjectMeta: api.ObjectMeta{
+					Name:      "test-" + hostname,
+					UID:       "12345",
+					Namespace: kubelet.NamespaceDefault,
+					SelfLink:  "/api/v1beta2/pods/test-" + hostname + "?namespace=default",
+				},
+				Spec: api.PodSpec{Containers: []api.Container{{Image: "test/image"}}},
+			}),
+		},
 	}
-}
-
-func TestReadManifestV1Beta2FromFile(t *testing.T) {
-	hostname, _ := os.Hostname()
-	hostname = strings.ToLower(hostname)
-
-	file := writeTestFile(t, os.TempDir(), "test_pod_config",
-		`{
-			"version": "v1beta2",
-			"uuid": "12345",
-			"id": "test",
-			"containers": [{ "name": "image", "image": "test/image", imagePullPolicy: "PullAlways"}]
-		}`)
-	defer os.Remove(file.Name())
-
-	ch := make(chan interface{})
-	NewSourceFile(file.Name(), time.Millisecond, ch)
-	select {
-	case got := <-ch:
-		update := got.(kubelet.PodUpdate)
-		expected := CreatePodUpdate(kubelet.SET, kubelet.FileSource, api.Pod{
-			ObjectMeta: api.ObjectMeta{
-				Name:      "test-" + hostname,
-				UID:       "12345",
-				Namespace: kubelet.NamespaceDefault,
-				SelfLink:  "/api/v1beta2/pods/test-" + hostname + "?namespace=default",
-			},
-			Spec: api.PodSpec{Containers: []api.Container{{Image: "test/image"}}},
-		})
-
-		if !api.Semantic.DeepDerivative(expected, update) {
-			t.Fatalf("Expected %#v, Got %#v", expected, update)
-		}
 
-	case <-time.After(time.Second):
-		t.Errorf("Expected update, timeout instead")
+	for _, testCase := range testCases {
+		func() {
+			file := writeTestFile(t, os.TempDir(), "test_pod_config", testCase.fileContents)
+			defer os.Remove(file.Name())
+
+			ch := make(chan interface{})
+			NewSourceFile(file.Name(), time.Millisecond, ch)
+			select {
+			case got := <-ch:
+				update := got.(kubelet.PodUpdate)
+				if !api.Semantic.DeepDerivative(testCase.expected, update) {
```

Comment:
DeepDerivative is used to get around all the default value population in the test. We should avoid using it if possible. In this case, it's justified though :)